### PR TITLE
Adding SQL Edge project template

### PR DESCRIPTION
**Pull request by Benjin**
This PR adds `projectTypeId` to the parameters passed to ProjectProviders when they're creating a new project so that they can differentiate between multiple project types/templates contributed.

It also adds a second project template for Edge projects, which should build/validate/deploy (but not work, 'cause the actual Edge hub streams wouldn't be set up) as-is, without any additional changes by the user.

image
image

### CODE REVIEW

File: extensions/sql-database-projects/src/common/constants.ts [hunk 1]
```diff
@@ -23,9 +23,13 @@ export const MicrosoftDatatoolsSchemaSqlSql = 'Microsoft.Data.Tools.Schema.Sql.S
...
-export const sqlDatabaseProjectTypeId = 'sqldbproj';
-export const projectTypeDisplayName = localize('projectTypeDisplayName', "SQL Database");
-export const projectTypeDescription = localize('projectTypeDescription', "Design, edit, and publish schemas for SQL databases");
+export const emptySqlDatabaseProjectTypeId = 'EmptySqlDbProj';
+export const emptyProjectTypeDisplayName = localize('projectTypeDisplayName', "SQL Database");
+export const emptyProjectTypeDescription = localize('projectTypeDescription', "Design, edit, and publish schemas for SQL databases from scratch");
+
+export const edgeSqlDatabaseProjectTypeId = 'SqlDbEdgeProj';
+export const edgeProjectTypeDisplayName = localize('edgeProjectTypeDisplayName', "SQL Database Edge");
+export const edgeProjectTypeDescription = localize('edgeProjectTypeDescription', "Start with the core pieces ready to go for SQL Edge databases");
```
Comment (Benjin) [inline extensions/sql-database-projects/src/common/constants.ts:32]: Replace strings/name with those from mockups.  Likely will need to revert "empty" description as well.

File: extensions/sql-database-projects/src/common/constants.ts [hunk 2]
```diff
@@ -23,9 +23,13 @@ export const MicrosoftDatatoolsSchemaSqlSql = 'Microsoft.Data.Tools.Schema.Sql.S
 export const databaseSchemaProvider = 'DatabaseSchemaProvider';
 
 // Project Provider
-export const sqlDatabaseProjectTypeId = 'sqldbproj';
-export const projectTypeDisplayName = localize('projectTypeDisplayName', "SQL Database");
-export const projectTypeDescription = localize('projectTypeDescription', "Design, edit, and publish schemas for SQL databases");
+export const emptySqlDatabaseProjectTypeId = 'EmptySqlDbProj';
+export const emptyProjectTypeDisplayName = localize('projectTypeDisplayName', "SQL Database");
```
Comment (Benjin) [inline extensions/sql-database-projects/src/common/constants.ts:27]: Update renamed `localize()` variable name

File: extensions/sql-database-projects/src/projectProvider/projectProvider.ts [hunk 3]
```diff
@@ -44,10 +44,17 @@ export class SqlDatabaseProjectProvider implements dataworkspace.IProjectProvide
...
+			displayName: constants.emptyProjectTypeDisplayName,
+			description: constants.emptyProjectTypeDescription,
+			icon: IconPathHelper.colorfulSqlProject
+		},
+		{
+			id: constants.edgeSqlDatabaseProjectTypeId,
+			projectFileExtension: constants.sqlprojExtension.replace(/\./g, ''),
+			displayName: constants.edgeProjectTypeDisplayName,
+			description: constants.edgeProjectTypeDescription,
 			icon: IconPathHelper.colorfulSqlProject
```
Comment (Benjin) [inline extensions/sql-database-projects/src/projectProvider/projectProvider.ts:58]: Replace with Edge project icon

File: extensions/sql-database-projects/src/templates/templates.ts [hunk 4]
```diff
@@ -52,11 +55,14 @@ export async function loadTemplates(templateFolderPath: string) {
 		loadObjectTypeInfo(storedProcedure, constants.storedProcedureFriendlyName, templateFolderPath, 'newTsqlStoredProcedureTemplate.sql'),
 		loadObjectTypeInfo(preDeployScript, constants.preDeployScriptFriendlyName, templateFolderPath, 'newTsqlPreDeployScriptTemplate.sql'),
 		loadObjectTypeInfo(postDeployScript, constants.postDeployScriptFriendlyName, templateFolderPath, 'newTsqlPostDeployScriptTemplate.sql'),
+		loadObjectTypeInfo(dataSource, constants.dataSource, templateFolderPath, 'newTsqlDataSourceTemplate.sql'),
```
Comment (Benjin) [inline extensions/sql-database-projects/src/templates/templates.ts:58]: Add these to command palette dropdown list (but probably not context menu)

**Review by Benjin — COMMENTED** [review]
TODO: add tests for multiple projects (data workspaces) and Edge template (db projects)

File: extensions/sql-database-projects/src/common/constants.ts [hunk 5]
```diff
@@ -23,9 +23,13 @@ export const MicrosoftDatatoolsSchemaSqlSql = 'Microsoft.Data.Tools.Schema.Sql.S
...
+export const emptyProjectTypeDescription = localize('projectTypeDescription', "Design, edit, and publish schemas for SQL databases from scratch");
```
Comment (kisantia) [inline extensions/sql-database-projects/src/common/constants.ts:28]: check with Drew before changing the description with this and for the Edge project description

Comment (Benjin) [reply extensions/sql-database-projects/src/common/constants.ts:28]: @dzsquared any ideas these descriptions?  I think we want strings that call out the difference between the one that starts from a blank state and the one that starts from a template, so I took a first pass.

Comment (dzsquared) [reply extensions/sql-database-projects/src/common/constants.ts:28]: @Benjin 
How about?

SQL Database:
Design, edit, and publish schemas for SQL databases starting from an empty project
SQL Edge:  (drop Database please)
Design, edit, and publish schemas for SQL Edge starting from some common Edge objects


<alternatively>
SQL Database:
Develop and publish schemas for SQL databases starting from an empty project
SQL Edge:  (drop Database please)
Start with the core pieces to develop and publish schemas for SQL Edge


my strong opinions are:
- avoiding database in SQL Edge
- avoiding "from scratch" to be non-colloquial

Comment (Benjin) [reply extensions/sql-database-projects/src/common/constants.ts:28]: Noted!  I like the second option better, since it leads with the fact that it's seeding the project with core pieces.

**Review by dzsquared — APPROVED** [review]

File: extensions/sql-database-projects/resources/templates/newTsqlExternalStreamingJobTemplate.sql [hunk 6]
```diff
@@ -5,6 +5,5 @@
 
 EXEC sys.sp_create_streaming_job @NAME = '@@OBJECT_NAME@@', @STATEMENT = 'INSERT INTO SqlOutputStream SELECT
```
Comment (udeeshagautam) [inline extensions/sql-database-projects/resources/templates/newTsqlExternalStreamingJobTemplate.sql:6]: two quick things,
1. please check that after replacng the @@vars with values the project builds/deploys fine
2.  when using references like SqlOutputStream  - check once how View template refers to Table - and lets be consistent on that pattern...

Comment (udeeshagautam) [reply extensions/sql-database-projects/resources/templates/newTsqlExternalStreamingJobTemplate.sql:6]: One more nit - the file names in the snapshot above are the name that you entered or that we get by default? - if default - we should have them consistent too. Lets have suffix 1 on all or none

**Review by udeeshagautam — APPROVED** [review]
looks great.

File: extensions/sql-database-projects/src/common/constants.ts (hunk 5 above)
Comment (kisantia) [reply extensions/sql-database-projects/src/common/constants.ts]: can you also update the screenshot in the PR description with the updated template name and descriptions?

**Review by kisantia — APPROVED** [review]

File: extensions/sql-database-projects/resources/templates/newTsqlExternalStreamingJobTemplate.sql (hunk 6 above)
Comment (Benjin) [reply extensions/sql-database-projects/resources/templates/newTsqlExternalStreamingJobTemplate.sql:6]: 1. Yup, the template builds, validates, and deploys without any changes from the user.

2. Sorry, what pattern of how we do references?

3. Agreed, and fixed (see updated screenshot)

**Comment by udeeshagautam** [comment]
@Benjin - we can merge this now right?
